// @tarojs/components/dist/cjs/taro-canvas-core.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-e751dfe3.js');

const indexCss = "taro-canvas-core{display:block;position:relative;width:300px;height:150px}";

const LONG_TAP_DELAY = 500;
const Canvas = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.onLongTap = index.createEvent(this, "longtap", 7);
    this.onTouchStart = () => {
      this.timer = setTimeout(() => {
        this.onLongTap.emit();
      }, LONG_TAP_DELAY);
    };
    this.onTouchMove = () => {
      clearTimeout(this.timer);
    };
    this.onTouchEnd = () => {
      clearTimeout(this.timer);
    };
    this.canvasId = undefined;
    this.nativeProps = {};
  }
  render() {
    const { canvasId, nativeProps } = this;
    return (index.h("canvas", Object.assign({ "canvas-id": canvasId, style: {
        width: '100%',
        height: '100%'
      }, onTouchStart: this.onTouchStart, onTouchMove: this.onTouchMove, onTouchEnd: this.onTouchEnd }, nativeProps)));
  }
  get el() { return index.getElement(this); }
};
Canvas.style = indexCss;

exports.taro_canvas_core = Canvas;
