// @tarojs/helper/dist/swcRegister.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InjectDefineConfigHeader = void 0;
const Visitor_js_1 = require("@swc/core/Visitor.js");
class InjectDefineConfigHeader extends Visitor_js_1.Visitor {
    visitTsType(expression) {
        return expression;
    }
    visitCallExpression(expression) {
        const callee = expression.callee;
        if (callee.type === 'Identifier' && (callee.value === 'definePageConfig' || callee.value === 'defineAppConfig')) {
            return expression.arguments[0].expression;
        }
        return expression;
    }
}
exports.InjectDefineConfigHeader = InjectDefineConfigHeader;
function createSwcRegister({ only, plugins }) {
    const config = {
        only: Array.from(new Set([...only])),
        jsc: {
            parser: {
                syntax: 'typescript',
                decorators: true
            },
            transform: {
                legacyDecorator: true
            }
        },
        module: {
            type: 'commonjs'
        }
    };
    if (plugins) {
        config.jsc.experimental = {
            plugins
        };
    }
    require('@swc/register')(config);
}
exports.default = createSwcRegister;
